File: Bitwise/Hard/LET_no1178_Number of Valid Words for Each Puzzle.py
class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        wTable = collections.defaultdict(int)
        for w in words:
            wTable[self.getMask(w)] += 1
        
        res = [0] * len(puzzles)
        for i in range(len(puzzles)):
            subMask = mask = self.getMask(puzzles[i])
            firstChar = ord(puzzles[i][0]) - ord('a')
            
            while subMask:
                if subMask >> firstChar & 1:
                    res[i] += wTable[subMask]
                subMask = (subMask - 1) & mask
        
        return res
    
    def getMask(self, w):
        res = 0
        for c in w:
            res |= (1 << (ord(c) - ord('a')))
        return res

File: Bitwise/Hard/LET_no1178_Number of Valid Words for Each Puzzle.cpp
class Solution {
public:
    vector<int> findNumOfValidWords(vector<string>& words, vector<string>& puzzles) {
        unordered_map<int, int> table;
        for (string& w : words) {
            int mask = getMask(w);
            table[mask]++;
        }
        
        int n = words.size();
        vector<int> res;
        for (string& w : puzzles) {
            int mask = getMask(w);
            int firstChar = 1 << (w[0] - 'a');
            int cur = 0;
            
            for (int sub = mask; sub > 0; sub = ((sub - 1) & mask)) {
                if ((sub & firstChar) && table.count(sub)) {
                    auto it = table.find(sub);
                    if (it != table.end())
                        cur += it->second;
                }
            }
            res.push_back(cur);
        }
        
        return res;
    }
    
    int getMask(string& w) {
        int res = 0;
        for (char c : w)
            res |= (1 << (c - 'a'));
        return res;
    }
};